File: 121 best time to buy and sell stock/BuyAndSellStock.java

public class BuyAndSellStock {
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length == 0) {
            return 0;
        }
        int max = prices[0];
        int min = prices[0];
        int res = 0;
        for (int price : prices) {
            if (price < min) {
                min = price;
                max = price;
            }
            if (price > max) {
                max = price;
            }
            res = Math.max(res, max - min);
        }
        return res;
    }
}

File: 121 best time to buy and sell stock/BuyAndSellStock.py
class BuyAndSellStock(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if not prices:
            return 0
        maxPrice = prices[0]
        minPrice = prices[0]
        res = 0
        
        for i in range(len(prices)):
            if prices[i] > maxPrice:
                maxPrice = prices[i]
            
            if prices[i] < minPrice:
                minPrice = prices[i]
                maxPrice = prices[i]
            
            
            res = max(res, maxPrice - minPrice)
        
        return res
        